Move MultiAppPostprocessorNearestElementTransfer from Pronghorn to MOOSE

YaqiWang opened the pull request:
Close #20913.

No changes are made in the source files, markdown files, and tests except the issue number is changed from  Pronghorn Issue 128 to MOOSE #20913.

YaqiWang committed:
move MultiAppPostprocessorNearestElementTransfer from Pronghorn to MOOSE framework #20913

YaqiWang committed:
changes for moving into moose #20914

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/1032060/) on 50e3aa9 : invalidated by @YaqiWang

pronghorn fixed.

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1032059/) on 50e3aa9 : invalidated by @YaqiWang

pronghorn fixed

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1032056/) on 50e3aa9 : invalidated by @YaqiWang

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1032068/) on 50e3aa9 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>5dbd36</th>
      <th colspan=4>#20914 50e3aa</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.51%</td>
      <td>82.52%</td>
      <td>+0.00%</td>
      <td>93.94%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>74713</td>
      <td>74772</td>
      <td>+59</td>
      <td>62</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15833</td>
      <td>15842</td>
      <td>+9</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20914/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20914/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20914/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad reviewed (dismissed):
I'll push the patch, but with moving this to the framework, we're going to do a some-to-all communication with `push_parallel_vector_data` for the postprocessor values instead of all-to-all

GiudGiud reviewed (changes requested):
This is the reverse direction of MultiAppVariableValuePostprocessorTransfer
https://mooseframework.inl.gov/source/transfers/MultiAppVariableValueSamplePostprocessorTransfer.html

it should be implemented as such

fdkong commented on framework/src/transfers/MultiAppPostprocessorNearestElementTransfer.C:
```diff
@@ -0,0 +1,136 @@
+
+  // Gather all the multiapps postprocessor values.
+  _communicator.min(pp_values);
+  _communicator.max(duplicate);
```
We should grab only the necessary values not everything

fdkong commented on framework/src/transfers/MultiAppPostprocessorNearestElementTransfer.C:
```diff
@@ -0,0 +1,136 @@
+    {
+      dof_id_type dof = elem->dof_number(_to_sys_num, _to_var_num, 0);
+      auto v = solution(dof);
+      Real nv = v * (1 - _relax_factor) + pp_values[_cached_multiapp_pos_ids[i]] * _relax_factor;
```
This is extremely hacking. I would like to see transfer only does transfer.


May separate it  into a pluggable system

fdkong replied:
I might not understand the motivation here

fdkong reviewed (dismissed):
A few comments

lindsayad committed:
Remove relaxation work that was not in Pronghorn

lindsayad committed:
Only do second all-to-all communication in debug mode

lindsayad commented:
@YaqiWang @fdkong I removed one of the all-to-all communications for all but `dbg` mode. I don't think I'll change to use `push/pull_parallel_vector_data` in this PR because I have some other fires that I should put out right now. @fdkong if you want to take the time to do that on this PR definitely feel free to push up a commit. Otherwise I am fine with the one all-to-all for now

YaqiWang commented:
@lindsayad you were not in the morning meeting. I think we agreed to have the relaxation factor in for a moment before https://github.com/idaholab/moose/issues/20100 can be addressed.

GiudGiud commented:
I agreed but Fande was against.
I spent all day on auxvariable relaxation, should be ready soon

YaqiWang committed:
merge the two transfers MultiAppPostprocessorNearestElementTransfer and MultiAppPostprocessorNearestElementTransfer #20913

YaqiWang committed:
merge the two test folders and remove two duplicated input files with cli_args #20913

YaqiWang committed:
update the markdown files and change an inaccurate requirement #20913

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1034111/) on 6692275 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20914/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20914/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a0ea072203682af7765d96ad3cdef64ecfc7e0e`

YaqiWang commented:
We possibly can remove this
```
  if (_directions.size() != 1)
    paramError("direction", "This transfer is only unidirectional");
```
But I am not sure, so I did not do it.

YaqiWang commented:
This transfer really should be based on an extra element ID, which is used for indexing sub-applications which will make the from-multiapp transfer well-defined. Using distance for doing the transfer may have some behavior sensitive to round-off. I will add an error message in the `initialSetup` to prevent this from happening.

YaqiWang committed:
add an error message to avoid unexpected behavior #20913

YaqiWang committed:
make sure the from transfer work with array variables and some trivial regoldings due to added new array variable #20913

YaqiWang committed:
rearrange tests a little #20913

YaqiWang committed:
remove deprecation in tutorials noticed due to changes in #20913

YaqiWang committed:
replace the if check with _directions instead of querying the parameters directly because users may be using the deprecated parameters #20913

lindsayad committed:
Use pull_parallel_vector_data instead of all-to-all during execute

Also inherit from MeshChangedInterface and make sure we re-setup
the element to postprocessor data caches after the mesh changes

lindsayad commented:
@fdkong let me know if this last commit looks good to you. If not, we can revert to the all-to-all during `execute`

YaqiWang commented:
@fdkong and @GiudGiud please review this. Thanks.

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -41,15 +47,104 @@ MultiAppVariableValueSamplePostprocessorTransfer::validParams()
     paramError("direction", "This transfer is only unidirectional");
 
-  if (isParamValid("from_multi_app"))
-    paramError("from_multi_app", "This transfer direction has not been implemented");
```
we still need an error for the sibling case since it was not implemented?

YaqiWang replied:
Is the above check `if (_directions.size() != 1)` covering this? I assume you mean the transfers among siblings.

GiudGiud replied:
sibling would be one direction still, between_multiapp

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -41,15 +47,104 @@ MultiAppVariableValueSamplePostprocessorTransfer::validParams()
+    if (!_fe_problem.getAuxiliarySystem().hasVariable(_var_name))
+      paramError("source_variable", "Variable must be an auxiliary variable");
+  }
+}
```
i feel like this transfer should check the execution flags of the postprocessors. 
If you dont want to do it, let me know and i ll create an issue

YaqiWang replied:
Not for this moment.

GiudGiud replied:
created #20969

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -41,15 +47,104 @@ MultiAppVariableValueSamplePostprocessorTransfer::validParams()
+          ++count;
+      }
+      if (count > 0)
+        mooseError("The distances of an element to more than one sub-applications are too close."
```
```suggestion
        mooseWarning("The distances of an element to more than one sub-applications are too close."
```

YaqiWang replied:
I prefer using mooseError more often than mooseWarning. Undetermined behavior should be considered as an error. Do you have arguments on making this as a warning?

GiudGiud replied:
it's not undetermined, you pick the element with the lowest ID. 

The argument is this needs to work for rapid prototyping and sometimes we dont care about this scenario as much as getting something done

YaqiWang replied:
Not a good argument. I'd rather the users be aware potential issues with more time on fixing the models and instead later complaining the code is not working as expected. The error message should give them enough information to check their models.

GiudGiud replied:
A warning gives them enough information.
If they dont want to heed it, that's their problem.

An error here breaks this transfer for entire classes of simulations.
Imagine you want to use a centroid multiapp but transfer onto a nodal variable?
Or you run a complex simulation with adaptivity and it develops this condition after 10k CPU hours?
Or the transition from one subapp to another is actually the correct behavior as the mesh is changing?

YaqiWang replied:
`transition from one subapp to another` won't hit this error. If starting with a CentroidMultiApp, mesh adaptation won't cause troubles. But I will make this as a warning.

GiudGiud replied:
it wont hit unless the middle point, the one that is undetermined, is on the way. 
CentroidMultiApp would hit for all nodal variables on a regular mesh

ok thx

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -41,15 +47,104 @@ MultiAppVariableValueSamplePostprocessorTransfer::validParams()
+      }
+      if (count > 0)
+        mooseError("The distances of an element to more than one sub-applications are too close."
+                   "\nDifferent positions for sub-applications or a centroid-based MultiApp can "
+                   "be used to resolve this error.");
```
```suggestion
                   "\nThe mapping from postprocessors to variables is not unique. The element with the lowest ID was chosen. Different positions for sub-applications or a centroid-based MultiApp can "
                   "be used to resolve this indetermination.");
```

GiudGiud replied:
a good candidate for a mooseWarningOnce but this isnt ready yet

YaqiWang replied:
I think there is a misunderstanding here. We have more candidate sub-apps providing the postprocessor values for an element. We must set values for all elements but now cannot choose one unless we choose the one with smaller sub-app id. I still think issuing an error is better than this bandage to make things deterministic. I can add here that users can choose to block restrict the variable to fix this.

GiudGiud replied:
smaller subapp id seems fine then. I must have missed that the for loop is on subapps not elements

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -41,15 +47,104 @@ MultiAppVariableValueSamplePostprocessorTransfer::validParams()
+    if (_var.hasBlocks(elem->subdomain_id()))
+    {
+      Real distance = std::numeric_limits<Real>::max();
+      unsigned int count = 0;
```
```suggestion
      unsigned int count = 0;
      unsigned int min_elem_id = std::numeric_limits<unsigned int>::max();
```

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -41,15 +47,104 @@ MultiAppVariableValueSamplePostprocessorTransfer::validParams()
+        Real current_distance = (getFromMultiApp()->position(j) - elem->true_centroid()).norm();
+        if (MooseUtils::absoluteFuzzyLessThan(current_distance, distance))
+        {
+          distance = current_distance;
+          multiapp_pos_id = j;
+          count = 0;
```
```suggestion
          distance = current_distance;
          if (elem->id() < min_elem_id)
          {
            multiapp_pos_id = j;
            min_elem_id = elem->id()
          }
          count = 0;
```

YaqiWang replied:
Why do we need this `min_elem_id`?

GiudGiud replied:
yeah we really need min_app_id not min_elem_id

lindsayad replied:
this implicitly uses the min app id since we are iterating over the number of apps and if distances are equal we use the first one we encounter. App ID is more deterministic than element ID ... element IDs may be affected by mesh distribution

GiudGiud commented on tutorials/darcy_thermo_mech/step10_multiapps/problems/step10.i:
```diff
@@ -192,17 +192,15 @@
     type = MultiAppPostprocessorInterpolationTransfer
-    direction = from_multiapp
-    multi_app = micro
+    from_multi_app = micro
```
thanks for catching those. there are inputs everywhere ugh

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -129,10 +219,95 @@ MultiAppVariableValueSamplePostprocessorTransfer::execute()
+          std::vector<dof_id_type> dof_indices;
+          _var.getDofIndices(elem, dof_indices);
+          mooseAssert(dof_indices.size() == 1,
+                      "The variable must be in constant monomial with one DoF on an element");
```
```suggestion
                      "The variable must be a constant monomial with one DoF on an element");
```

GiudGiud reviewed (changes requested):
does re-partitioning trigger meshChanged() btw?

lindsayad commented:
> does re-partitioning trigger meshChanged() btw?

`meshChanged` will be called when adaptivity is performed. Adaptivity typically results in a re-partitioning

GiudGiud commented:
ok so common use case is caught and for the narrow one I dont think we do that anywhere else than what I tried in Macaw so i think we are good on this question

YaqiWang committed:
addressing comments from Guillaume #20913

YaqiWang committed:
add a check on not doing sibling transfer #20913

YaqiWang committed:
relax an error into a warning #20913

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -129,12 +129,15 @@ MultiAppVariableValueSamplePostprocessorTransfer::cacheElemToPostprocessorData()
+        mooseWarning("The distances of an element to more than one sub-applications are too close "
+                     " in transfer '",
+                     name(),
+                     "'.\nHaving different positions for sub-applications, "
```
need to tell user results might be affected

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:
```diff
@@ -129,12 +129,15 @@ MultiAppVariableValueSamplePostprocessorTransfer::cacheElemToPostprocessorData()
+                     name(),
+                     "'.\nHaving different positions for sub-applications, "
+                     "a centroid-based MultiApp or adding block restriction to the variable can "
+                     "be used to resolve this error.");
```
```suggestion
                     "be used to resolve this warning.");
```
